app/code/Magento/InventoryCatalog/Test/Integration/DeleteLegacyCatalogInventoryPluginTest.php

```diff
@@ -0,0 +1,125 @@
+     */
+    private $stockItemCriteriaFactory;
+
+    /**
+     * @var ProductRepositoryInterface
+     */
+    private $productRepository;
+
+    /**
+     * @var SourceItemRepositoryInterface
+     */
+    private $sourceItemRepository;
+
+    /**
+     * @var SearchCriteriaBuilder
+     */
+    private $searchCriteriaBuilder;
+
+    /**
+     * @var SourceItemsDeleteInterface
+     */
+    private $sourceItemsDelete;
+
+    /**
+     * @var StockRegistryInterface
+     */
+    private $stockRegistry;
+
+    /**
+     * @var DefaultSourceProviderInterface
+     */
+    private $defaultSourceProvider;
+
+    protected function setUp()
+    {
+        $this->oldStockItemRepository = Bootstrap::getObjectManager()->get(StockItemRepositoryInterface::class);
+        $this->stockItemCriteriaFactory = Bootstrap::getObjectManager()->get(StockItemCriteriaInterfaceFactory::class);
+
+        $this->productRepository = Bootstrap::getObjectManager()->get(ProductRepositoryInterface::class);
+
+        $this->sourceItemRepository = Bootstrap::getObjectManager()->get(SourceItemRepositoryInterface::class);
+        $this->searchCriteriaBuilder = Bootstrap::getObjectManager()->create(SearchCriteriaBuilder::class);
+
+        $this->sourceItemsDelete = Bootstrap::getObjectManager()->get(SourceItemsDeleteInterface::class);
+        $this->stockRegistry = Bootstrap::getObjectManager()->get(StockRegistryInterface::class);
+        $this->defaultSourceProvider = Bootstrap::getObjectManager()->get(DefaultSourceProviderInterface::class);
+    }
+
+    /**
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/products.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/sources.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/stocks.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/source_items.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/stock_source_link.php
+     * @magentoDbIsolation enabled
+     */
+    public function testDeleteDefaultSourceItemTriggersDeleteStockItem()
+    {
+        /** @var Product $product */
+        $productSku = 'SKU-4';
+        $product = $this->productRepository->get($productSku);
+
+        /** @var StockItemCriteriaInterface  $criteria */
+        $defaultStock = $this->stockRegistry->getStock();
+        $criteria = $this->stockItemCriteriaFactory->create();
+        $criteria->setProductsFilter([$product->getId()]);
+        $criteria->setStockFilter($defaultStock);
+        $criteria->addFilter('filter_is_in_stock', StockItemInterface::IS_IN_STOCK, true);
+
+        /** @var StockItemCollectionInterface $collectionBeforeChange */
+        $stockItemsBeforeDelete = $this->oldStockItemRepository->getList($criteria)->getItems();
+        $this->assertCount(1, $stockItemsBeforeDelete);
+
+        $searchCriteria = $this->searchCriteriaBuilder
```

Comment:
Looks like we have a few times with code
Maybe it will be better to create sugar service (not interface but implementation) and put under hood this logic (I mean creating and configuring of Search Criteria)